### DISCUSSION TITLE
[@types/react-table] How to set type when passing props

### DISCUSSION THREAD

**Discussion by decayedCell** [DefinitelyTyped/DefinitelyTyped#60185] [Issues with a @types package]
Hello. 

I am using a type file similar to [this](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/react-table) and I have no issues creating my table
``` 
const {
    getTableProps, getTableBodyProps, headerGroups, prepareRow, page, canPreviousPage,
    canNextPage, pageOptions, pageCount, gotoPage, nextPage, previousPage, setPageSize,
    setGlobalFilter, setFilter, setAllFilters, toggleHideColumn,
    state: {
      globalFilter, filters, pageIndex, pageSize,
    },
  } = useTable(
    {
      columns,
      data,
      initialState: {
        pageSize: 10,
        pageIndex: 0,
      },
    },
    useGlobalFilter,
    useFilters,
    useSortBy,
    usePagination,
  );
```

etc.

My issue is when I try to pass some of those functions/properties to a child component.
From parent:
```
<ChildComponent
        nextPage={nextPage}
        previousPage={previousPage}
        gotoPage={gotoPage}
        setPageSize={setPageSize}
        setGlobalFilter={setGlobalFilter}
        canPreviousPage={canPreviousPage}
        canNextPage={canNextPage}
        pageIndex={pageIndex}
        pageOptions={pageOptions}
        pageCount={pageCount}
        pageSize={pageSize}
        globalFilter={globalFilter}
      />
```
To child:
```
function ChildComponent(props: any): JSX.Element {
  const {
    nextPage, previousPage, gotoPage, setPageSize, setGlobalFilter,
    canPreviousPage, canNextPage, pageIndex, pageOptions, pageCount,
    globalFilter, pageSize,
  } = props;
```

How do I type check props properly in this instance? If I don't use :any I get the typical 'Paramater 'props' implicitly has an 'any' type'

[a comment by an automated account is not shown]

**Comment by ggascoigne** [comment]
I think that https://github.com/ggascoigne/react-table-example/blob/master/src/Table/TablePagination.tsx#L88 might be similar to what you are trying to do.

**Comment by decayedCell** [comment]
Interesting, but I think my inexperience with Typescript won't let me find the solution yet. I rewrote it like this:
```
function ChildComponent<T extends Record<string, unknown>>({
  instance,
}: PropsWithChildren<{ instance: TableInstance<T> }>): JSX.Element {
const {
    state: {
      globalFilter, pageIndex, pageSize,
    },
    nextPage, previousPage, gotoPage, setPageSize, setGlobalFilter,
    canPreviousPage, canNextPage, pageOptions, pageCount,
  } = instance;
```
This seems to be nice as VS Code shows the correct type on them now and no errors,
but there are errors on passing the prop from Parent
```
<ChildComponent
        nextPage={nextPage}
```

```
[{
	"message": "Type '{ nextPage: () => void; previousPage: () => void; gotoPage: (updater: number | ((pageIndex: number) => number)) => void; setPageSize: (pageSize: number) => void; setGlobalFilter: (filterValue: any) => void; ... 6 more ...; globalFilter: any; }' is not assignable to type 'IntrinsicAttributes & { instance: TableInstance<Record<string, unknown>>; } & { children?: ReactNode; }'.\n  Property 'nextPage' does not exist on type 'IntrinsicAttributes & { instance: TableInstance<Record<string, unknown>>; } & { children?: ReactNode; }'.",
}]
```

**Reply by ggascoigne** [reply]
Ah, so your issue is more of a React and TypeScript one, as opposed to a react-table one, I should have read your question a bit more carefully.  Sorry about that.

There's a mismatch between how you are defining `ChildComponent` and how you appear to want to use it.

Going back to your first example (since as I said, I might have sent you off in slightly the wrong direction).

Your definition is:

```tsx
function ChildComponent(props: any): JSX.Element {
  const {
    nextPage, previousPage, gotoPage, setPageSize, setGlobalFilter,
    canPreviousPage, canNextPage, pageIndex, pageOptions, pageCount,
    globalFilter, pageSize,
  } = props;
```

That is really suggesting that props is an object that can be destructured as though it was a table instance.

To declare it like that, you'd write:

```tsx
function ChildComponent<T extends Record<string, unknown>>( props: TableInstance<T>): JSX.Element {
  const {
    nextPage, previousPage, gotoPage, setPageSize, setGlobalFilter,
    canPreviousPage, canNextPage, pageIndex, pageOptions, pageCount,
    globalFilter, pageSize,
  } = props;
```

That defines props as being a `TableInstance<T>` and will  let you use it like:

```tsx
<ChildComponent
        nextPage={nextPage}
        previousPage={previousPage}
        gotoPage={gotoPage}
        setPageSize={setPageSize}
        setGlobalFilter={setGlobalFilter}
        canPreviousPage={canPreviousPage}
        canNextPage={canNextPage}
        pageIndex={pageIndex}
        pageOptions={pageOptions}
        pageCount={pageCount}
        pageSize={pageSize}
        globalFilter={globalFilter}
      />
```

At least I think it will, not tested it.

The other approach which I tend to prefer is to explicitly pass the instance in as an instance property, and that is more in line with the second definition you posted.

```tsx
function ChildComponent<T extends Record<string, unknown>>({
  instance,
}: { instance: TableInstance<T> }): JSX.Element {
const {
    state: {
      globalFilter, pageIndex, pageSize,
    },
    nextPage, previousPage, gotoPage, setPageSize, setGlobalFilter,
    canPreviousPage, canNextPage, pageOptions, pageCount,
  } = instance;
```

Note that using `PropsWithChildren` is only necessary if you are actually passing children to  the Component (I might have done it in that example, but I've learnt better since then and need to update it), 

Anyway, this second definition declares that `ChildComponent` is a component that gets one param, called instance that is a TableInstance.

To you use you need to grab the instance in the first place:

```tsx
const instance = useTable(
    {
      columns,
      data,
      initialState: {
        pageSize: 10,
        pageIndex: 0,
      },
    },
    useGlobalFilter,
    useFilters,
    useSortBy,
    usePagination,
  );

const {
    getTableProps, getTableBodyProps, headerGroups, prepareRow, page, canPreviousPage,
    canNextPage, pageOptions, pageCount, gotoPage, nextPage, previousPage, setPageSize,
    setGlobalFilter, setFilter, setAllFilters, toggleHideColumn,
    state: {
      globalFilter, filters, pageIndex, pageSize,
    },
  } = instance;

```

Then to use it you'd just do:

```tsx
  <ChildComponent
    instance={instance>
  />
```

Hope that's clear and help.

**Reply by decayedCell** [reply]
The first example I already tried before, the 'issue' is with it is that Typescript expects every single property that I wrote into TableInstance. This kind of makes sense, but even making a smaller TableInstance just for those doesn't work for me. UsePaginationOptions<D> for example, includes pageCount which I need but also includes things like 'paginateExpandedRows' which I don't use yet so it expects me to send that prop too. I would need to export a interface from 'react-table-config.d.ts' that uses exactly only the props I'm using. This may be possible, but I don't know how to do it yet.

As for the second example, it worked, but I had to change to 
``` 
function ChildComponent({
  instance,
}: { instance: TableInstance<object> }): JSX.Element {
```
Thank you for your help, I think I am happy with this version and what I learned, looking forward to v8.

**Reply by ggascoigne** [reply]
Oh, if you want a type that is like another type, but ever parameter is optional then take a look at using `Partial<>` https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype.  Either way, glad you got things working.
